src/business-logic/services/rundown-lock-service.ts

```diff
@@ -0,0 +1,125 @@
+import { InTransition } from '../../model/value-objects/in-transition'
+import { RundownService } from './interfaces/rundown-service'
+
+const RUNDOWN_LOCK_INTERVAL = 500
+const RUNDOWN_LOCK_ERROR_TEXT = 'Unable to do action. An action was already executed less than 500ms ago'
+
+export class RundownLockService implements RundownService {
+  private static instance: RundownService
+  private hasEnoughTimePassed: boolean = true
+
+  constructor(private readonly rundownService: RundownService) {}
+
+  public static getInstance(rundownService: RundownService): RundownService {
+    if (!this.instance) {
+      this.instance = new RundownLockService(rundownService)
+    }
+    return this.instance
+  }
+
+  public takeNext(rundownId: string): Promise<void> {
+    if (!this.hasEnoughTimePassed) {
+      throw new Error(RUNDOWN_LOCK_ERROR_TEXT)
+    }
+    this.hasEnoughTimePassed = false
+    setTimeout(() => (this.hasEnoughTimePassed = true), RUNDOWN_LOCK_INTERVAL)
+    return this.rundownService.takeNext(rundownId)
+  }
+
+  public deleteRundown(rundownId: string): Promise<void> {
+    if (!this.hasEnoughTimePassed) {
+      throw new Error(RUNDOWN_LOCK_ERROR_TEXT)
+    }
+    this.hasEnoughTimePassed = false
+    setTimeout(() => (this.hasEnoughTimePassed = true), RUNDOWN_LOCK_INTERVAL)
+    return this.rundownService.deactivateRundown(rundownId)
+  }
+
+  public activateRundown(rundownId: string): Promise<void> {
+    if (!this.hasEnoughTimePassed) {
+      throw new Error(RUNDOWN_LOCK_ERROR_TEXT)
+    }
+    this.hasEnoughTimePassed = false
+    setTimeout(() => (this.hasEnoughTimePassed = true), RUNDOWN_LOCK_INTERVAL)
+    return this.rundownService.activateRundown(rundownId)
+  }
+
+  public enterRehearsal(rundownId: string): Promise<void> {
+    if (!this.hasEnoughTimePassed) {
+      throw new Error(RUNDOWN_LOCK_ERROR_TEXT)
+    }
+    this.hasEnoughTimePassed = false
+    setTimeout(() => (this.hasEnoughTimePassed = true), RUNDOWN_LOCK_INTERVAL)
+    return this.rundownService.enterRehearsal(rundownId)
+  }
+
+  public deactivateRundown(rundownId: string): Promise<void> {
+    return this.rundownService.deactivateRundown(rundownId)
+  }
+
+  public setNext(rundownId: string, segmentId: string, partId: string, owner?: Owner): Promise<void> {
+    if (!this.hasEnoughTimePassed) {
+      throw new Error(RUNDOWN_LOCK_ERROR_TEXT)
+    }
+    this.hasEnoughTimePassed = false
+    setTimeout(() => (this.hasEnoughTimePassed = true), RUNDOWN_LOCK_INTERVAL)
+    return this.rundownService.setNext(rundownId, segmentId, partId, owner)
+  }
+
+  public resetRundown(rundownId: string): Promise<void> {
+    if (!this.hasEnoughTimePassed) {
+      throw new Error(RUNDOWN_LOCK_ERROR_TEXT)
+    }
+    this.hasEnoughTimePassed = false
+    setTimeout(() => (this.hasEnoughTimePassed = true), RUNDOWN_LOCK_INTERVAL)
+    return this.rundownService.resetRundown(rundownId)
+  }
+
+  public insertPartAsOnAir(rundownId: string, part: Part): Promise<void> {
```

Comment:
`insertPartAsOnAir`, `insertPartAsNext`, `insertPieceAsNext`, `insertPieceAsOnAir` and `replacePieceOnAirOnNextPart` are not supposed to be behind the lock.
Read the acceptance criteria for the things that should be behind the lock.

Reply:
Fix